### PR TITLE
chore: Tidy up buttons (fix #2517)

### LINKED ISSUE #2517: Refactor Button component, mixin, and colours
The Photon colours are a bit different than what's in the Sketch mock and our Button mixin is a bit more verbose than it needs to be: we should clean it up.

See: https://github.com/mozilla/addons-frontend/pull/2478#pullrequestreview-42061544 and https://github.com/mozilla/addons-frontend/pull/2478#pullrequestreview-42062630

### LINKED ISSUE #3705: "Register/Log in" button needs to be made solid
We don't have a ghost/outline button spec yet in the design system, so until then, we need to make this button solid. I'll provide the proper spec for this soon.
screen shot 2017-10-30 at 7 42 34 pm

### LINKED ISSUE #3752: Fallback buttons in the disco pane regressed with the "+"
No-one will generally see this, but if you load the disco pane and mozAddonManager is disabled it looks like this:

discover_add-ons

### PR DESCRIPTION
**Pull request by tofumatt**
Fix #2517 
Fix #3752 
Fix #3705

Nearly finished and I need to post my screenshots, but this updates the buttons to be in line with the mocks.

### Disco Pane
With `mozAddonManager` unavailable, see #3752.
screenshot 2018-01-26 15 38 22

### Screenshots
screenshot 2018-01-25 23 41 16
screenshot 2018-01-25 23 31 45
screenshot 2018-01-25 23 31 18
screenshot 2018-01-25 23 31 13
screenshot 2018-01-25 23 31 08
screenshot 2018-01-25 23 30 55
screenshot 2018-01-25 23 30 22
screenshot 2018-01-25 23 21 00
screenshot 2018-01-25 23 20 48

### CODE REVIEW

**Comment by tofumatt** [comment]
I've been at this on and off for a few days and I'm about to head off on PTO so this might be a bit rushed, but I wanted to post what I have. I think it might actually be all set to go but it's one of those things I've been working on a bit here-and-there for long enough that I need another set of eyes 😄

File: src/core/components/Paginate/Paginate.scss [hunk 1]
```diff
@@ -18,16 +18,14 @@
...
-    $background-active: $light-active-color,
-    $background-hover: $light-hover-color,
-    $color: $link-color
-  );
+  // @include button(
+  //   $background: $light-base-color,
+  //   $background-active: $light-active-color,
+  //   $background-hover: $light-hover-color,
+  //   $color: $link-color
+  // );
```
Comment (willdurand) [inline src/core/components/Paginate/Paginate.scss]: 😛

File: src/ui/components/Button/index.js [hunk 2]
```diff
@@ -16,6 +27,8 @@ export default class Button extends React.Component {
     disabled: PropTypes.boolean,
     href: PropTypes.string,
     to: PropTypes.string,
+    type: PropTypes.string,
```
Comment (willdurand) [inline src/ui/components/Button/index.js]: I see many `type=none` already, I guess this could be a good default value in `defaultProps`, what do you think?

File: src/ui/components/Button/index.js [hunk 3]
```diff
@@ -24,12 +37,22 @@ export default class Button extends React.Component {
...
 
+    if (!type || !BUTTON_TYPES.includes(type)) {
+      throw new Error(`"${type}" supplied but that is not a valid button type`);
+    }
+
     const setClassName = (...classConfig) => {
-      return classNames('Button', className, ...classConfig);
+      return classNames(
+        'Button', `Button--${type}`, className, ...classConfig, {
+          width: `Button--width-${width}`,
```
Comment (willdurand) [inline src/ui/components/Button/index.js]: I dont really like this freestyle here. I think it would be nice to have a `boolean` prop called `fullwidth` (default: `false`) and if set, then we add the `Button--widthfull` or `Button-fullwidth` class.

File: src/ui/components/Button/index.js [hunk 4]
```diff
@@ -6,8 +6,19 @@ import PropTypes from 'prop-types';
...
+
+const BUTTON_TYPES = [
+  'neutral',
+  'light',
+  'action',
+  'cancel',
+  'confirm',
+  'alert',
+  'report',
+  'none',
```
Comment (willdurand) [inline src/ui/components/Button/index.js]: It is not sorted, isn't it? 🤓

File: tests/unit/core/components/TestPaginate.js [hunk 5]
```diff
@@ -191,54 +176,53 @@ describe('<Paginate />', () => {
...
+  //     expect(links.map((link) => [link.textContent, link.getAttribute('href')])).toEqual([
+  //       ['Previous', '/some-path/?page=4'],
+  //       ['3', '/some-path/?page=3'],
+  //       ['4', '/some-path/?page=4'],
+  //       ['6', '/some-path/?page=6'],
+  //       ['7', '/some-path/?page=7'],
+  //       ['Next', '/some-path/?page=6'],
+  //     ]);
+  //   });
+  // });
```
Comment (willdurand) [inline tests/unit/core/components/TestPaginate.js]: 😛

File: tests/unit/ui/components/TestButton.js [hunk 6]
```diff
@@ -34,7 +34,7 @@ describe('<Button />', () => {
   it('renders a link with a href', () => {
     const href = 'https://addons.mozilla.org';
     const button = shallow(
-      <Button className="Bar" href={href}>Link text!</Button>);
+      <Button className="Bar" href={href} type="none">Link text!</Button>);
```
Comment (willdurand) [inline tests/unit/ui/components/TestButton.js]: ah, `none` means `link` like... `<a>`?

**Comment by muffinresearch** [comment]
It would be good to check with @pwalm but this suggests that non-radiused buttons should still be on the disco pane - see https://github.com/mozilla/addons-frontend/issues/3060 for details.

**Comment by pwalm** [comment]
# Button tidying master post!
## Design and specs
There are three sizes of buttons:
- Default
- Micro
- Puffy

There are three types/styles:
- Primary
- Default
- Ghost

[The spec is here](http://design.firefox.com/photon/components/buttons.html) and is very clear.

## Buttons
screen shot 2017-11-16 at 3 02 54 pm

screen shot 2017-11-16 at 3 03 05 pm

screen shot 2017-11-16 at 3 03 18 pm

screen shot 2017-11-16 at 3 03 27 pm

screen shot 2017-11-16 at 3 03 40 pm

screen shot 2017-11-16 at 3 03 44 pm

screen shot 2017-11-16 at 3 05 38 pm

screen shot 2017-11-16 at 3 05 56 pm

There will prob be some edge cases! We'll address 'em as we discover 'em.

File: tests/unit/ui/components/TestButton.js (hunk 6 above)
Comment (tofumatt) [reply tests/unit/ui/components/TestButton.js]: I'm late to this response, but yes. It's there for a few things we want to use mostly unstyled buttons for.

**Comment by tofumatt** [comment]
Hey @kumar303,

This one is finally ready for review. I've removed some unneeded paginator tests that I caught while refactoring buttons... I forget how they came up but those code paths are covered elsewhere now and this PR is from Nov so 🤷‍♂️

This refactors the `<Button>` API and styles according to the Photon spec.

File: src/core/css/inc/mixins.scss [hunk 7]
```diff
@@ -105,199 +97,6 @@
...
-
-  .Icon {
-    @include margin-end(6px);
-
-    height: 16px;
-    vertical-align: top;
-    width: 16px;
-  }
-
-  &.Button--small {
```
Comment (kumar303) [inline src/core/css/inc/mixins.scss]: This is still referenced in the Notice styles: https://github.com/mozilla/addons-frontend/blob/67da4111618ba06c0967e461929174597aeba7e6/src/ui/components/Notice/styles.scss#L44

Maybe it just needs to be renamed in `Notice/styles.scss`?

File: src/ui/css/vars.scss [hunk 8]
```diff
@@ -53,9 +53,13 @@ $font-size-header-title: 34px;
 
 // Border radius sizes
 $border-radius-default: 9px;
-$border-radius-s: 4px;
-$border-radius-m: 8px;
+$border-radius-s: 3px;
```
Comment (kumar303) [inline src/ui/css/vars.scss]: I think it's dangerous that we have variables that are named the same as those in https://github.com/mozilla/addons-frontend/blob/67da4111618ba06c0967e461929174597aeba7e6/src/core/css/inc/vars.scss#L35-L37

...because Sass won't even tell you when an import overwrites variables from a previous import. Could we call these `$ui-border-radius-`, maybe?

Also, why change `$border-radius-s` to 3px? Was that only for buttons? My concern is that we use `$border-radius-s` all over the place and the Sketch file still has it at 4px for most things. But, again, maybe changing the variable names would address my concern.

File: tests/unit/core/components/TestPaginate.js [hunk 9]
```diff
@@ -27,49 +19,42 @@ describe('<Paginate />', () => {
       ...getRenderProps(),
       ...extra,
     };
-    return findRenderedComponentWithType(renderIntoDocument(
-      <Paginate {...props} />
-    ), Paginate).getWrappedInstance();
+
+    return shallowUntilTarget(<Paginate {...props} />, PaginateBase);
```
Comment (kumar303) [inline tests/unit/core/components/TestPaginate.js]: ❤️

File: tests/unit/core/components/TestPaginate.js [hunk 10]
```diff
@@ -191,54 +176,11 @@ describe('<Paginate />', () => {
...
-          </Router>
-        ), node, () => {
-          resolve(node);
-        });
-      });
-    }
-
-    return renderPaginateRoute().then((root) => {
-      const links = Array.from(root.querySelectorAll('a'));
-      expect(links.map((link) => [link.textContent, link.getAttribute('href')])).toEqual([
```
Comment (kumar303) [inline tests/unit/core/components/TestPaginate.js]: Did this get moved somewhere? I know it wasn't a great test, just making sure you deleted it on purpose.

File: tests/unit/ui/components/TestButton.js [hunk 11]
```diff
@@ -7,10 +7,18 @@ import { createFakeEvent } from 'tests/unit/helpers';
 
 
 describe('<Button />', () => {
```
Comment (kumar303) [inline tests/unit/ui/components/TestButton.js]: This could be `describe('__filename', ...` while you're in here.

File: src/ui/components/Button/index.js [hunk 12]
```diff
@@ -6,30 +6,62 @@ import PropTypes from 'prop-types';
...
-import './Button.scss';
+
+const BUTTON_TYPES = [
+  'neutral',
+  'light',
+  'action',
+  'cancel',
+  'confirm',
+  'alert',
+  'none',
```
Comment (kumar303) [inline src/ui/components/Button/index.js]: Is this really necessary? Can it just be considered `none` if `type` is undefined?

Comment (tofumatt) [reply src/ui/components/Button/index.js]: Yeah, that's fair. This started to make sure I caught any errant buttons in the conversion that didn't have a type. I think going forward an empty type meaning none is fine 👍

File: src/ui/components/Button/index.js [hunk 13]
```diff
@@ -6,30 +6,62 @@ import PropTypes from 'prop-types';
...
 export default class Button extends React.Component {
   static propTypes = {
     children: PropTypes.node,
     className: PropTypes.string,
     disabled: PropTypes.boolean,
     href: PropTypes.string,
+    micro: PropTypes.boolean,
+    puffy: PropTypes.boolean,
     to: PropTypes.string,
+    type: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/ui/components/Button/index.js]: Ouch, yes, that's a good call re: the `type` prop in `...rest`... I thought that as soon as you pointed out this for renaming. 😄

Easier grepping is good too. Totally for that; I'll go through and rename the prop.

File: src/ui/components/Button/styles.scss [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+        width: $micro-font-size;
+      }
+
+      &:active:not(.Button--disabled) {
+        padding-top: 1px;
+      }
+
+      // Small text is difficult to read against the bright Photon colors.
+      // This isn't (yet) part of the spec but @tofumatt added it because
+      // he thinks it improves accessibility.
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/ui/components/Button/styles.scss]: Yes, I did that with a notification bar earlier and meant to do that here... must've forgot. Will do!

File: tests/unit/core/components/TestPaginate.js (hunk 10 above)
Comment (tofumatt) [reply tests/unit/core/components/TestPaginate.js]: I checked and those code paths are covered in the `PaginateLink` tests and in other tests in this file. They're just leftovers from a much earlier time and they aren't needed anymore.

File: src/ui/css/vars.scss (hunk 8 above)
Comment (tofumatt) [reply src/ui/css/vars.scss]: I'd like to change the variable names; that's a good point. Changing the size though was seemingly unrelated and I've undone it. But I think it's worth doing in another patch as this is unrelated 😄

**Comment by tofumatt** [comment]
I've made all the changes requested here, I think it's good for another look.

File: src/ui/components/Button/styles.scss [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+  transition-timing-function: $easing-curve;
+
+  .Icon {
+    @include margin-end(6px);
+
+    height: $default-font-size;
+    width: $default-font-size;
+  }
+
+  // We frustratingly target :link directly so we need this override.
```
Comment (kumar303) [inline src/ui/components/Button/styles.scss]: I thought one of the goals for refactoring was to try using cascades instead of mixins so that these overrides wouldn't be needed. Did cascades not work out?

**Comment by kumar303** [comment]
buttons in the `ErrorList` look good:

screenshot 2018-01-30 16 13 30

File: src/ui/components/Button/styles.scss (hunk 15 above)
Comment (tofumatt) [reply src/ui/components/Button/styles.scss]: Oh wait, this is because we target link styles using `:link` which messes with usual cascades and such... it's gonna have to stay for now (this was here before, just moved), but yeah until we can nix https://github.com/mozilla/addons-frontend/blob/master/src/amo/components/Addon/styles.scss it's gotta stay.

I'd like to do that but it should be it's own issue as there's a large possibility of regressing link styles.
